evidence/pool.go

```diff
@@ -140,32 +139,139 @@ func (evpool *Pool) AddEvidence(ev types.Evidence) error {
-	return VerifyEvidence(evidence, evpool.State(), evpool.stateDB, evpool.blockStore)
+// CheckEvidence takes an array of evidence from a block and verifies all the evidence there.
+// If it has already verified the evidence then it jumps to the next one. It ensures that no
+// evidence has already been committed or is being proposed twice. It also adds any
+// evidence that it doesn't currently have so that it can quickly form ABCI Evidence later.
+// Check evidence returns the hash of the entire evidence list so that it can be compares with
+// the headers EvidenceHash.
+func (evpool *Pool) CheckEvidence(evList types.EvidenceList) error {
+	hashes := make([][]byte, len(evList))
+	for idx, ev := range evList {
+
+		ok := evpool.fastCheck(ev)
+
+		if !ok {
+			evInfo, err := evpool.verify(ev)
+			if err != nil {
+				return &types.ErrEvidenceInvalid{Evidence: ev, Reason: err}
+			}
+
+			if err := evpool.addPendingEvidence(evInfo); err != nil {
+				evpool.logger.Error("Can't add evidence to pending list", "err", err, "evInfo", evInfo)
+			}
+
+			evpool.logger.Info("Verified new evidence of byzantine behavior", "evidence", ev)
+		}
+
+		// check for duplicate evidence. We cache hashes so we don't have to work them out again.
+		hashes[idx] = ev.Hash()
+		for i := idx - 1; i >= 0; i-- {
+			if bytes.Equal(hashes[i], hashes[idx]) {
+				return &types.ErrEvidenceInvalid{Evidence: ev, Reason: errors.New("duplicate evidence")}
+			}
+		}
+	}
+
+	return nil
 }
 
-// MarkEvidenceAsCommitted marks all the evidence as committed and removes it
-// from the queue.
-func (evpool *Pool) MarkEvidenceAsCommitted(height int64, evidence []types.Evidence) {
+// ABCIEvidence processes all the evidence in the block, marking it as committed and removing it
+// from the pending database. It then forms the individual abci evidence that will be passed back to
+// the application.
+func (evpool *Pool) ABCIEvidence(height int64, evidence []types.Evidence) []abci.Evidence {
 	// make a map of committed evidence to remove from the clist
 	blockEvidenceMap := make(map[string]struct{})
+	abciEvidence := make([]abci.Evidence, 0)
 	for _, ev := range evidence {
+
+		// get entire evidence info from pending list
+		infoBytes, err := evpool.evidenceStore.Get(keyPending(ev))
+		if err != nil {
+			evpool.logger.Error("Unable to retrieve evidence to pass to ABCI. "+
+				"Evidence pool should have seen this evidence before",
+				"evidence", ev, "err", err)
+			continue
+		}
+		var infoProto evproto.Info
+		err = infoProto.Unmarshal(infoBytes)
+		if err != nil {
+			evpool.logger.Error("Decoding evidence info failed", "err", err, "height", ev.Height(), "hash", ev.Hash())
+			continue
+		}
+		evInfo, err := InfoFromProto(&infoProto)
+		if err != nil {
+			evpool.logger.Error("Converting evidence info from proto failed", "err", err, "height", ev.Height(),
+				"hash", ev.Hash())
+			continue
+		}
+		abciEv := make([]abci.Evidence, len(evInfo.Validators))
```

Comment:
why we need two slices here: abciEvidence and abciEv?

Reply:
Yeah I can just create one. It's just with one slice we know the size and the other slice adds these slices together so we don't know the size of that. But I can just use the append function for them all

Reply:
yeah, you can set capacity make([]abci.Evidence, 0, len(evInfo.Validators))

there's also cap(X) functions which returns capacity